main/src/library/Map.flix

```diff
@@ -142,19 +158,61 @@ namespace Map {
     ///
     /// Returns `None` if `m` is empty.
     ///
-    pub def minimum(m: Map[k, v]): Option[(k, v)] =
-        let Map(xs) = m;
-        RedBlackTree.minimum(xs)
+    pub def minimumKey(m: Map[k, v]): Option[(k, v)] =
+        let Map(t) = m;
+        RedBlackTree.minimum(t)
 
     ///
     /// Optionally finds `k => v` where `k` is the smallest key according to the given comparator `cmp`.
     ///
     /// Returns `None` if `m` is empty.
     ///
+    /// Purity polymorphic: Runs in parallel when given a pure function `cmp`.
+    ///
     @ParallelWhenPure
-    pub def minimumBy(cmp: (k, v, k, v) -> Comparison & ef, m: Map[k, v]): Option[(k, v)] & ef =
-        let Map(t) = m;
-        RedBlackTree.minimumBy(cmp, t)
+    pub def minimumKeyBy(cmp: (k, k) -> Comparison & ef, m: Map[k, v]): Option[(k, v)] & ef =
+        let f = r -> cmp(r.kl, r.kr);
+        reifyEff(f) {
+            case Pure(g) =>
+                if (useParallelEvaluation(m))
+                    let h = (kl, _, kr, _) -> g({kl = kl, kr = kr});
+                    let Map(t) = m;
+                    RedBlackTree.parMinimumBy(threads() - 1, h, t) as & Pure
+                else
+                    reduceLeftWithKey((kl, vl, kr, vr) -> if (cmp(kl, kr) == LessThan) (kl, vl) else (kr, vr), m)
```

Comment:
I just had an idea:

Introduce a private function called `min(k1, v1, k2, v2)` which is:
```
reduceLeftWithKey((kl, vl, kr, vr) -> if (cmp(kl, kr) == LessThan) (kl, vl) else (kr, vr), m)
```

and use it everywhere. Then I think things become a lot cleaner.